src/nwb_conversion_tools/tools/hdmf.py

```diff
@@ -1,7 +1,77 @@
+"""Collection of modifications of HDMF functions that are to be tested/used on this repo until propagation upstream."""
 from typing import Tuple
 
 import numpy as np
-from hdmf.data_utils import GenericDataChunkIterator
+from hdmf.data_utils import GenericDataChunkIterator as HDMFGenericDataChunkIterator
+
+
+class GenericDataChunkIterator(HDMFGenericDataChunkIterator):
+    def _get_default_buffer_shape(self, buffer_gb: float = 1.0) -> Tuple[int]:
+        chunk_bytes = np.prod(self.chunk_shape) * self.dtype.itemsize
+        assert buffer_gb > 0, f"buffer_gb ({buffer_gb}) must be greater than zero!"
+        assert (
+            buffer_gb >= chunk_bytes / 1e9
+        ), f"buffer_gb ({buffer_gb}) must be greater than the chunk size ({chunk_bytes / 1e9})!"
+        assert all(
+            np.array(self.chunk_shape) > 0
+        ), f"Some dimensions of chunk_shape ({self.chunk_shape}) are less than zero!"
+
+        maxshape = np.array(self.maxshape)
+
+        # Early termination condition
+        if np.prod(maxshape) * self.dtype.itemsize / 1e9 < buffer_gb:
+            return tuple(self.maxshape)
+
+        buffer_bytes = chunk_bytes
+        axis_sizes_bytes = maxshape * self.dtype.itemsize
+        smallest_chunk_axis, second_smallest_chunk_axis, *_ = np.argsort(self.chunk_shape)
+        target_buffer_bytes = buffer_gb * 1e9
+
+        # If the smallest full axis does not fit within the buffer size, form a square along the two smallest axes
+        sub_square_buffer_shape = np.array(self.chunk_shape)
+        if min(axis_sizes_bytes) > target_buffer_bytes:
+            k1 = np.floor((target_buffer_bytes / chunk_bytes) ** 0.5)
+            for axis in [smallest_chunk_axis, second_smallest_chunk_axis]:
+                sub_square_buffer_shape[axis] = k1 * sub_square_buffer_shape[axis]
+            return tuple(sub_square_buffer_shape)
+
+        # Original one-shot estimation has good performance for certain shapes
+        k2 = np.floor(
+            (buffer_gb * 1e9 / (np.prod(self.chunk_shape) * self.dtype.itemsize)) ** (1 / len(self.chunk_shape))
+        )
+        unpadded_buffer_shape = tuple(
+            [
+                min(max(int(x), self.chunk_shape[j]), self.maxshape[j])
+                for j, x in enumerate(k2 * np.array(self.chunk_shape))
+            ]
+        )
+        unpadded_buffer_bytes = np.prod(unpadded_buffer_shape) * self.dtype.itemsize
```

Comment:
I think you don't need those parentheses there. I think you can do tuple comprehension directly without them.

Also, up to you, but I would do two list comprehension here for clarity / readability. One that calculates the max between equi-distributed shape and another that clips over max shape. But I think you like brevity more than I do.  Along those lines, you could also use something like `np.clip` to sort this out as well in one sweep I think (to clip equi-distributed shapes between max_shapes and chunk_shapes.

Reply:
You mean separate compressions over the min vs. the max?

Reply:
`np.clip` is nice - thanks for pointing that out!